### PR TITLE
For blurry icons for HiDPI

### PR DESCRIPTION
**Pull request by MoetaYuko**
Most of this PR serves as clean-ups so that a comprehensive regression test might be desired.

Before:
swappy-20230421_154351

After:
swappy-20230421_154125

### CODE REVIEW

**Comment by nwg-piotr** [comment]
Please DO READ the [general contributing rules for the nwg-shell project](https://nwg-piotr.github.io/nwg-shell/contribution).

**Comment by MoetaYuko** [comment]
> Please DO READ the [general contributing rules for the nwg-shell project](https://nwg-piotr.github.io/nwg-shell/contribution).

Yes, I have read them before making the PR. I guess your main concerns are:
> Make sure you don’t introduce new bugs. We already have enough bugs to fix.
Make sure your work is finished and needs no futher effort to be usable.

I have tried my best to test all routines being touched here, except for `hyprland_taskbar.py` which I don't use, but its logic is almost the same as `sway_taskbar.py` so it *should* be fine. I'm just not 100% sure that no regression is introduced as I may fail to cover some corner cases.

**Comment by nwg-piotr** [comment]
Sure. What I mean is: I am a single, hobbyist, after hours developer. There's no team to analyze, test and review complex PRs. Just me, currently working on nwg-dock-hypraland, with about 10 various PRs waiting for me to understand them, test and merge. Your work is quite large, so I'd expect at least a brief introduction: what the changes are about, what changes you did to the code. Otherwise your PR may have to wait a long time for me to able to take care of it.

**Review by MoetaYuko — COMMENTED** [review]
Summary:
1. Observing the original `create_pixbuf` always fallbacks to `icons_light/icon-missing.svg`, it's moved to the outmost try-except block so we no longer need to duplicate it 3 times.
2. `if icon_name.startswith("/")`, then load from the absolute path (same as before). Otherwise, there are 3 routines in order if `icons_path` supplied:
    1.  If `icons_path` supplied, load from `icons_path/icon_name.svg`
    2. load icon_name from icon theme
    3. convert icon_name to lowercase and load from icon theme

    and if `icons_path` is not supplied, skip i. In short, ii. and iii. are executed when 
Previously, ii. and iii. were duplicated in i.'s if and else blocks. This commit avoids duplication by putting ii. and iii. in i.'s except block,

**Comment by nwg-piotr** [comment]
Thank you very much!

File: nwg_panel/tools.py [hunk 1]
```diff
@@ -601,79 +601,68 @@ def player_metadata():
...
+
+
+def update_gtk_entry(entry, icon_pos, icon_name, icon_size, icons_path=""):
+    scale = entry.get_scale_factor()
+    icon_size *= scale
+    pixbuf = create_pixbuf(icon_name, icon_size, icons_path)
+    entry.set_icon_from_pixbuf(icon_pos, pixbuf)
+
+
+def create_pixbuf(icon_name, icon_size, icons_path="", fallback=True):
```
Comment (MoetaYuko) [inline nwg_panel/tools.py:634]: Summary of changes:
1. Observing the original `create_pixbuf` always fallbacks to `icons_light/icon-missing.svg`, it's moved to the outmost try-except block so we no longer need to duplicate it 3 times.
2. `if icon_name.startswith("/")`, then load from the absolute path (same as before). Otherwise, there are 3 routines in order if `icons_path` supplied:
    1.  If `icons_path` supplied, load from `icons_path/icon_name.svg`
    2. load icon_name from icon theme
    3. convert icon_name to lowercase and load from icon theme

    and if `icons_path` is not supplied, skip i. In short, ii. and iii. are executed when 
Previously, ii. and iii. were duplicated in i.'s if and else blocks. This commit avoids duplication by putting ii. and iii. in i.'s except block.

The added `fallback` option raises an exception when the icon is not found, rather than fallback to `icon-missing.svg`. This enables the caller to do custom error handling.

File: nwg_panel/modules/executor.py [hunk 2]
```diff
@@ -81,55 +82,46 @@ def __init__(self, settings, icons_path, executor_name):
             Gdk.threads_add_timeout_seconds(GLib.PRIORITY_LOW, settings["interval"], self.refresh)
 
     def update_widget(self, output):
```
Comment (MoetaYuko) [inline nwg_panel/modules/executor.py:84]: The length of `output` can be either 1 or 2:
1. contains either icon label or icon path
2. a tuple of `(icon_path, label)`

The aims here are to fill in corresponding widgets and hide the non-existent one when `len(output) == 1`.
The logic was a mess and looks much better now.

File: nwg_panel/modules/hyprland_taskbar.py [hunk 3]
```diff
@@ -162,38 +162,11 @@ def __init__(self, settings, client, position, icons_path):
...
-                        try:
-                            pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(icon_from_desktop,
-                                                                            self.settings["image-size"],
-                                                                            self.settings["image-size"])
-                        except:
-                            pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(
-                                os.path.join(get_config_dir(), "icons_light/icon-missing.svg"),
-                                self.settings["image-size"],
-                                self.settings["image-size"])
-                        image.set_from_pixbuf(pixbuf)
```
Comment (MoetaYuko) [inline nwg_panel/modules/hyprland_taskbar.py:195]: This part additionally considers the image named `icon_from_desktop` compared to `update_image`, so a new helper function `update_image_fallback_desktop` is added to match it when `update_image` fails.

File: nwg_panel/modules/sni_system_tray/tray.py [hunk 4]
```diff
@@ -29,37 +29,15 @@ def resize_pix_buf(image, pixbuf, icon_size):
...
-    image.set_from_pixbuf(pixbuf)
-
-
-def load_icon(image, icon_name: str, icon_size, icons_path=""):
-    icon_theme = Gtk.IconTheme.get_default()
-    search_path = icon_theme.get_search_path()
-    try:
-        if icons_path:
-            search_path.append(icons_path)
-            icon_theme.set_search_path(search_path)
```
Comment (MoetaYuko) [inline nwg_panel/modules/sni_system_tray/tray.py:41]: `load_icon` did almost the same as `create_pixbuf` except for this search path stuff. `set_search_path` is added to `create_pixbuf` now, so we simply call it rather than duplicate the logic.

File: nwg_panel/modules/sni_system_tray/tray.py [hunk 5]
```diff
@@ -29,37 +29,15 @@ def resize_pix_buf(image, pixbuf, icon_size):
...
-            "tray -> update_icon: icon not found\n  icon_name: {}\n  search_path: {}".format(
-                icon_name,
-                search_path
-            ),
-            file=sys.stderr
-        )"""
-        path = os.path.join(get_config_dir(), "icons_light/icon-missing.svg")
-        pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(path, icon_size, icon_size)
+    scale = image.get_scale_factor()
+    surface = Gdk.cairo_surface_create_from_pixbuf(pixbuf, scale, image.get_window())
```
Comment (MoetaYuko) [inline nwg_panel/modules/sni_system_tray/tray.py:33]: key hidpi fix: apply scaling

File: nwg_panel/modules/sni_system_tray/tray.py [hunk 6]
```diff
@@ -29,37 +29,15 @@ def resize_pix_buf(image, pixbuf, icon_size):
...
+    image.set_from_surface(surface)
 
+
+def load_icon(image, icon_name: str, icon_size, icon_path=""):
+    scale = image.get_scale_factor()
+    icon_size *= scale
```
Comment (MoetaYuko) [inline nwg_panel/modules/sni_system_tray/tray.py:39]: key hidpi fix: load icon with higher resolution

File: nwg_panel/modules/sway_taskbar.py [hunk 7]
```diff
@@ -144,31 +144,7 @@ def __init__(self, tree, con, settings, position, icons_path, cache_file, floati
...
-                        try:
-                            pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(icon_from_desktop, settings["image-size"],
-                                                                            settings["image-size"])
-                        except:
-                            pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(
-                                os.path.join(get_config_dir(), "icons_light/icon-missing.svg"), settings["image-size"],
-                                settings["image-size"])
-                        image.set_from_pixbuf(pixbuf)
-
+            update_image_fallback_desktop(image, name, settings["image-size"], icons_path)
```
Comment (MoetaYuko) [inline nwg_panel/modules/sway_taskbar.py:147]: same as explained in hyprland_taskbar.py

File: nwg_panel/modules/sway_workspaces.py [hunk 8]
```diff
@@ -201,25 +202,22 @@ def refresh(self):
...
-                if "/" not in icon_from_desktop and not icon_from_desktop.endswith(
-                        ".svg") and not icon_from_desktop.endswith(".png"):
-                    update_image(self.icon, icon_from_desktop, self.settings["image-size"], self.icons_path)
-                else:
-                    pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(icon_from_desktop, self.settings["image-size"],
-                                                                    self.settings["image-size"])
-                    self.icon.set_from_pixbuf(pixbuf)
-            else:
-                image = Gtk.Image()
-                update_image(image, win_id, self.settings["image-size"], self.icons_path)
```
Comment (MoetaYuko) [inline nwg_panel/modules/sway_workspaces.py:216]: same as explained in hyprland_taskbar.py

File: nwg_panel/modules/sway_workspaces.py [hunk 9]
```diff
@@ -201,25 +202,22 @@ def refresh(self):
...
+            try:
+                update_image_fallback_desktop(self.icon,
+                                              win_id,
+                                              self.settings["image-size"],
+                                              self.icons_path,
+                                              fallback=False)
+                loaded_icon = True
```
Comment (MoetaYuko) [inline nwg_panel/modules/sway_workspaces.py:213]: mark success. won't reach this line if `update_image_fallback_desktop` raises an exception

File: nwg_panel/modules/sway_workspaces.py [hunk 10]
```diff
@@ -201,25 +202,22 @@ def refresh(self):
...
+            except:
+                pass
 
-            if not self.icon.get_visible():
-                self.icon.show()
-        else:
-            if self.icon.get_visible():
-                self.icon.hide()
+        if not loaded_icon and self.icon.get_visible():
+            self.icon.hide()
```
Comment (MoetaYuko) [inline nwg_panel/modules/sway_workspaces.py:220]: hide icon if failed to load, rather than showing a red cross.

File: nwg_panel/tools.py [hunk 11]
```diff
@@ -601,79 +601,68 @@ def player_metadata():
     return data
 
 
-def update_image(image, icon_name, icon_size, icons_path=""):
```
Comment (MoetaYuko) [inline nwg_panel/tools.py:604]: did exactly the same as `create_pixbuf`, drop copypasta

File: nwg_panel/tools.py [hunk 12]
```diff
@@ -601,79 +601,68 @@ def player_metadata():
...
-                except:
-                    path = os.path.join(get_config_dir(), "icons_light/icon-missing.svg")
-                    pixbuf = GdkPixbuf.Pixbuf.new_from_file_at_size(path, icon_size, icon_size)
-            if image:
-                image.set_from_pixbuf(pixbuf)
+def update_image(image, icon_name, icon_size, icons_path="", fallback=True):
+    scale = image.get_scale_factor()
+    icon_size *= scale
+    pixbuf = create_pixbuf(icon_name, icon_size, icons_path, fallback)
+    surface = Gdk.cairo_surface_create_from_pixbuf(pixbuf, scale, image.get_window())
```
Comment (MoetaYuko) [inline nwg_panel/tools.py:608]: key hidpi fix: apply scaling

File: nwg_panel/tools.py [hunk 13]
```diff
@@ -601,79 +601,68 @@ def player_metadata():
...
+    image.set_from_surface(surface)
 
 
-def create_pixbuf(icon_name, icon_size, icons_path=""):
-    # In case a full path was given
-    if icon_name.startswith("/"):
-        try:
+def update_image_fallback_desktop(image, icon_name, icon_size, icons_path, fallback=True):
```
Comment (MoetaYuko) [inline nwg_panel/tools.py:612]: same as explained in hyprland_taskbar.py

**Comment by MoetaYuko** [comment]
@nwg-piotr I think I have added necessary comments to help understand the changes, feel free to drop a note for confusing parts.

**Comment by nwg-piotr** [comment]
Thanks. I'll test it over the weekend.

**Comment by nwg-piotr** [comment]
It seems to work as expected. Thank you for the contribution.
